R/teal_data-class.R

```diff
@@ -0,0 +1,103 @@
+setOldClass("JoinKeys")
+
+#' Reproducible data class.
+#'
+#' Reproducible data class basing on [`teal.code::qenv-class`].
+#' Don't interact with slots directly, it is suggested to use methods instead
+#' (see `methods(class = "teal_data")`).
+#'
+#' @name teal_data-class
+#' @rdname teal_data-class
+#'
+#' @slot code (`character`) representing code necessary to reproduce the environment.
+#'  To extract the `code` please use [get_code()].
+#' @slot env (`environment`) environment which content was generated by the evaluation
+#'  of the `code` slot. To extract variables from the environment please use [get_var()] or [`[[`].
+#' @slot id (`integer`) random identifier of the code element to make sure uniqueness
+#'  when joining.
+#' @slot warnings (`character`) the warnings output when evaluating the code.
+#' To extract the `warnings` use [get_warnings()].
+#' @slot messages (`character`) the messages output when evaluating the code
+#' @slot join_keys (`JoinKeys`) object.
+#'  To extract the `join_keys` use [get_join_keys()].
+#' @slot datanames (`character`) names of datasets in `env`. Needed when non-dataset
+#'  objects are needed in the `env` slot.
+#'  To extract the `datanames` use [get_dataname()].
+#'
+#' @import teal.code
+#' @keywords internal
+setClass(
+  Class = "teal_data",
+  contains = "qenv",
+  slots = c(join_keys = "JoinKeys", datanames = "character"),
+  prototype = list(
+    join_keys = join_keys(),
+    datanames = character(0)
+  )
+)
+
+#' Initialize `teal_data` object
+#'
+#' Initialize `teal_data` object.
+#' @name new_teal_data
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Accepts and stores comments also.
+#' @param data (`named list`) List of data.
+#' @param keys (`JoinKeys`) object
+#' @param datanames (`character`) names of datasets passed to `data`.
+#'   Needed when non-dataset objects are needed in the `env` slot.
+#'
+#' @examples
+#' new_teal_data(data = list(a = 1), code = quote(a <- 1))
+#' new_teal_data(data = list(a = 1), code = parse(text = "a <- 1"))
+#' new_teal_data(data = list(a = 1), code = "a <- 1")
+#'
+#' @export
+setGeneric("new_teal_data", function(data = list(), code = character(), keys = join_keys(), datanames = character()) {
+  standardGeneric("new_teal_data")
+})
+
+#' @rdname new_teal_data
+#' @export
+setMethod(
+  "new_teal_data",
+  signature = c(data = "list", code = "character", keys = "ANY"),
+  function(data, code, keys = join_keys(), datanames = names(data)) {
+    new_env <- rlang::env_clone(list2env(data), parent = parent.env(.GlobalEnv))
+    lockEnvironment(new_env, bindings = TRUE)
```

Comment:
why it's needed? maybe worth to put it in the in-line comment?

Reply:
environment is locked to prevent making changes when somebody tries to modify `obj@env` directly instead of using `eval_code` or `within`. Don't know though if lock really works. I'll investigate this

Reply:
I've checked lockEnvironment. Works as expected and covered by tests here

https://github.com/insightsengineering/teal.data/blob/bb6e0ffc350b9939adb306ae8cdd71229a21b015/tests/testthat/test-teal_data.R#L86